File: src/main/java/com/pancisin/bookster/rest/controllers/UserController.java
package com.pancisin.bookster.rest.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.fasterxml.jackson.annotation.JsonView;
import com.pancisin.bookster.model.User;
import com.pancisin.bookster.models.views.Summary;
import com.pancisin.bookster.repository.PageRepository;
import com.pancisin.bookster.repository.UserRepository;

@RestController
@RequestMapping(value = "api/user/{user_id}")
public class UserController {

	@Autowired
	private UserRepository userRepository;

	@Autowired
	private PageRepository pageRepository;

	@JsonView(Summary.class)
	@GetMapping()
	public ResponseEntity<User> getUserData(@PathVariable Long user_id) {
		User user = userRepository.findOne(user_id);

		if (user != null)
			return ResponseEntity.ok(user);
		else
			return new ResponseEntity<User>(HttpStatus.NOT_FOUND);
	}

	@GetMapping("event")
	public ResponseEntity<?> getEvents(@PathVariable Long user_id) {
		User stored = userRepository.findOne(user_id);
		return ResponseEntity.ok(stored.getEvents());
	}

	@JsonView(Summary.class)
	@GetMapping("/page")
	public ResponseEntity<?> getPage(@PathVariable Long user_id) {
		return ResponseEntity.ok(pageRepository.getByOwner(user_id));
	}

	@GetMapping("/followed-pages")
	@JsonView(Summary.class)
	public ResponseEntity<?> getFollowedPages(@PathVariable Long user_id) {
		User stored = userRepository.findOne(user_id);
		return ResponseEntity.ok(stored.getFollowed());
	}

}

File: src/main/kotlin/com/pancisin/bookster/rest/controllers/v1/UserController.kt
package com.pancisin.bookster.rest.controllers.v1

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity

import com.fasterxml.jackson.annotation.JsonView
import com.pancisin.bookster.components.annotations.PrivacyRestricted
import com.pancisin.bookster.model.User
import com.pancisin.bookster.models.views.Summary
import com.pancisin.bookster.repository.EventRepository
import com.pancisin.bookster.repository.PageRepository
import com.pancisin.bookster.repository.UserRepository
import com.pancisin.bookster.repository.custom.impl.UserSearchRepository
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("api/v1/user/{user_id}")
class UserController {

  @Autowired
  lateinit var userRepository: UserRepository

  @Autowired
  lateinit var pageRepository: PageRepository

  @Autowired
  lateinit var eventRepository: EventRepository

  @Autowired
  lateinit var userSearchRepository: UserSearchRepository

  @JsonView(Summary::class)
  @GetMapping("/page")
  fun getPage(@PathVariable user_id: Long) = ResponseEntity.ok(pageRepository.getByOwner(user_id))

  @PrivacyRestricted(constraint = "followed-pages")
  @GetMapping("/followed-pages")
  @JsonView(Summary::class)
  fun getFollowedPages(@PathVariable user_id: Long) = ResponseEntity.ok(pageRepository.getFollowed(user_id))

  @GetMapping("/privacy-constraints")
  fun getPrivacyConstraints(@PathVariable user_id: Long) = ResponseEntity.ok(userRepository.findOne(user_id).privacyConstraints)
}
